devutils/update_lists.py

```diff
@@ -205,84 +205,106 @@ def _check_regex_match(file_path, search_regex):
+                       domain_sub_list, pruning_list, symlink_list):
+    """
+    Adds the path to appropriate lists to be used by compute_lists.
+
+    path is the pathlib.Path to the file from the current working directory.
+    used_pep_list is a list of PRUNING_EXCLUDE_PATTERNS that have been matched
+    used_pip_list is a list of PRUNING_INCLUDE_PATTERNS that have been matched
+    used_dep_list is a list of DOMAIN_EXCLUDE_PREFIXES that have been matched
+    used_dip_list is a list of DOMAIN_INCLUDE_PATTERNS that have been matched
+    domain_sub_list is a list of files that require domain substitution
+    pruning_list is a list of file to be pruned
+    symlink_list is a list of symlinks
+    """
+    if not path.is_file():
+        # NOTE: Path.rglob() does not traverse symlink dirs; no need for special handling
+        return
+    relative_path = path.relative_to(_source_tree)
+    if path.is_symlink():
+        try:
+            resolved_relative_posix = path.resolve().relative_to(_source_tree).as_posix()
+        except ValueError:
+            # Symlink leads out of the source tree
+            return
+        symlink_list.append((resolved_relative_posix, relative_path.as_posix()))
+        return
+    try:
+        if should_prune(path, relative_path, used_pep_list, used_pip_list):
+            pruning_list.append(relative_path.as_posix())
+        elif should_domain_substitute(path, relative_path, used_dep_list, used_dip_list):
+            domain_sub_list.append(relative_path.as_posix())
+    except: #pylint: disable=bare-except
+        get_logger().exception('Unhandled exception while processing %s', relative_path)
+
+
+def compute_lists():
     """
     Compute the binary pruning and domain substitution lists of the source tree.
-    Returns a tuple of two items in the following order:
+    Returns a tuple of three items in the following order:
     1. The sorted binary pruning list
     2. The sorted domain substitution list
-
-    source_tree is a pathlib.Path to the source tree
-    search_regex is a compiled regex object to search for domain names
+    3. An UnusedPatterns object
     """
-    pruning_set = set()
-    domain_substitution_set = set()
-    deferred_symlinks = dict() # POSIX resolved path -> set of POSIX symlink paths
-    source_tree = source_tree.resolve()
+    # Set up the multiprocess lists
+    manager = Manager()
+    used_pep_list = manager.list() # pruning_exclude_patterns
+    used_pip_list = manager.list() # pruning_include_patterns
+    used_dep_list = manager.list() # domain_exclude_prefixes
+    used_dip_list = manager.list() # domain_include_patterns
+    domain_sub_list = manager.list()
+    pruning_list = manager.list()
+    symlink_list = manager.list() # POSIX resolved path -> set of POSIX symlink paths
+
+    # Launch multiple processes iterating over the source tree
+    with Pool() as procpool:
+        procpool.starmap(
```

Comment:
I believe the chunksize is 1 by default. We should probably increase this because the IPC overhead might be impacting performance significantly. You could try playing with a few different values, but we don't have to be super rigorous about this.

We could also hardcode this value for now, unless you think it's useful enough to have as a CLI argument (I personally can't think of any use-cases where this would be useful enough. Seems like we'd be micro-optimizing for the particular system we're on at that point.)